coaster/sqlalchemy/properties.py

```diff
@@ -0,0 +1,197 @@
+        post.state.RECENT       # True (this one calls the validator if the base state matches)
+
+    States can also be used for database queries when accessed from the class::
+
+        # Generates MyPost._state == MY_STATE.DRAFT
+        MyPost.query.filter(*MyPost.state.DRAFT)
+
+        # Generates MyPost._state.in_(MY_STATE.UNPUBLISHED)
+        MyPost.query.filter(*MyPost.state.UNPUBLISHED)
+
+        # Generates MyPost._state == MY_STATE.PUBLISHED, MyPost.datetime > datetime.utcnow() - timedelta(hours=1))
+        MyPost.query.filter(*MyPost.state.RECENT)
+
+    Since added states with a validator have more than one condition that must match,
+    the class-level property returns a tuple of filter conditions.
+
+    States can be set by directly changing the attribute::
+
+        post.state = MY_STATE.PENDING
+        post.state = 'some_invalid_value'  # This will raise a ValueError
+
+    However, state change via the @transition decorator adds more power:
+
+    1. Original and final states can be specified, prohibiting a transition from any other state.
+    2. The transition method can do additional validation and housekeeping.
+
+    :param str propname: Name of the property that is to be wrapped
+    :param LabeledEnum lenum: The LabeledEnum containing valid values
+    :param str doc: Optional docstring
+    """
+    def __init__(self, propname, lenum, doc=None):
+        self.propname = propname
+        self.lenum = lenum
+        self.__doc__ = doc
+        self.states = {}
+
+    def __get__(self, obj, cls=None):
+        return _StatePropertyWrapper(self, obj, cls)
+
+    def __set__(self, obj, value):
+        if value not in self.lenum:
+            raise ValueError("Not a valid value: %s" % value)
+
+        type(obj).__dict__[self.propname].__set__(obj, value)
+
+    # Since __get__ never returns self, the following methods will only be available
+    # within the owning class's namespace. It will not be possible to call them outside
+    # the class to add additional states or transitions. If you really must do that,
+    # use cls.__dict__['state_property'].add_state, etc.
+
+    def add_state(self, name, value, validator, class_validator=None):
+        """
+        Add an additional state that combines an existing state with a validator
+        that must also pass. The validator receives the object on which the property
+        is present as a parameter.
+
+        :param str name: Name of the new state
+        :param value: Value or group of values of an existing state
+        :param validator: Function that will be called with the host object as a parameter
+        :param class_validator: Function that will be called when the state is queried
+            on the class instead of the instance. Falls back to ``validator`` if not specified
+        """
+        self.states[name] = (value, validator, class_validator)
+
+    def transition(self, from_, to):
```

Comment:
Is it necessary to mention `from_`? Would there be cases where `from_` may not matter?

Reply:
I see it allows a list, so I suppose that helps.